### ISSUE TITLE
Clients can't connect to oort Cloud

### ISSUE THREAD

**Issue by leopal77** [cometd/cometd#768]
We have the same setup mentioned in this [thread](https://github.com/cometd/cometd/issues/767) (setupB).

We have added a CometListener for `cometJoined` and `cometLeft` events to ensure that the 2 nodes of our setup have correctly identified each other. From the logs we can see that the connection between the two oort comets is established.

However, when we access the part of the app where cometd is involved , the browser(client) cannot connect to cometd server. We get a `402:Unkonow Client` as a response message for `connect` request. Just to mention that `handshake` request is successful. 

Some additional info about the setup. If one of the two cluster nodes is down the above misfunctionality is not taking place. So I think that the problem is probably a configuration for clustering that we have missed. 

In documentation is mentioned that if an app doesn't need to support message exchange between clients `Seti` is not required. In our app, clients can only subscribe to channels not directly publish to them. Cometd server publishes to channels where clients have subscribed so from my point of view  we do not need `Seti`. Do we?

**Comment by sbordet** [comment]
What do you have between the client and the server nodes to direct traffic ? A load balancer ?

You need `Seti` if you need to find a specific client in the cluster.
If an event arrives to `nodeB1` and that event if destined to a specific client that is connected to `nodeB2`, then you will need `Seti`.
If your events are always broadcast to add clients, you don't need `Seti`.

**Comment by leopal77** [comment]
It is clear that we don't need `Seti` as events are never destined to spesific client, only to channels that client have subscribed.

You are right, for directing traffic between nodes, we use a load balancer. Any advice on how I can track the source of the problem? I have added breakpoints to cometd source code but I can not understand much. For some reason, `processMetaConnect` method of `AbstractHttpTransport.java` keeps failing..

**Comment by sbordet** [comment]
For CometD to work, you need to make sure that the load balancer sticks each client to the same server, as explained in [this section](https://docs.cometd.org/current/reference/#_java_oort).

**Comment by leopal77** [comment]
This is my load balancer configuration. I tried setting `ProxySet stickysession=BAYEUX_BROWSER`  but no luck.

```Listen 9091
<VirtualHost *:9091>
        ServerName balancer_name
        ProxyRequests Off

        RedirectMatch ^$ /context/
        RedirectMatch ^/$ /context/
        <Location /context/>
            ProxyPass  balancer://balancer_name/context/
            ProxyPassReverse balancer://balancer_name/context/
        </Location>

        <Proxy balancer://balancer_name>
                BalancerMember http://node1IP:port route=node1 retry=10 smax=1 ttl=15 loadfactor=1
                BalancerMember http://node2IP:port route=node2 retry=10 smax=1 ttl=15 loadfactor=1
                ProxySet lbmethod=byrequests
                ProxySet stickysession=JSESSIONID
                ProxySet nofailover=Off
                ProxySet timeout=5
        </Proxy>
</VirtualHost>```

**Comment by sbordet** [comment]
Cannot help you here, as I don't know what load balancer is that. There are plenty of people using Oort behind a load balancer, so I am sure it is only matter for you to find the right configuration for your load balancer.

**Comment by leopal77** [comment]
It's an apache httpd load balancer. Is there any known tested working load balancer configuration to use for my tests?I need to eliminate the root of the problem and I think it's a good step.

 Another question, in `OortStaticConfigurationServlet` for `oort.url` and `oort.cloud` parameters should I use the ip addresses of nodes or their hostnames(DNS is used for hostname resolution)?

Thanks

**Comment by sbordet** [comment]
I'm not familiar with Apache httpd.

For `oort.url` and `oort.cloud` typically IP addresses are best, unless you also want the nodes to be resolved via DNS to leverage other load balancing mechanisms.

**Comment by leopal77** [comment]
Hello Simon,

I have made a simple project that can reproduce the misfunctionality I have reported in this thread.

You can check it out at the following [link](https://github.com/leopal77/CometdAuctionDemo).

I am using Wildfly(10.1.0 Final) as application server with default `standalone-full-ha.xml`. For load balancer, as I mentioned above, I use Apache httpd but I am pretty sure you can use anything that you are familiar with. It would be nice if you could test  with Wildfly as changing application server is unfortunately not feasible.

Thank you in advance .

**Comment by sbordet** [comment]
I would suggest another course of action.

You setup 2 nodes without load balancer.
You connect client1 explicitly to nodeA and client2 explicitly to nodeB.
You verify that your application works and that both client1 and client2 can correctly receive server events, no matter whether the server event arrives to nodeA or to nodeB.

If the above works, it means that it's not a CometD problem, nor a Wildfly problem, but just a load balancer problem.
If that's the case, you only need to correctly configure the load balancer.
If that's not the case, please report evidence of where the communication is lost.

**Comment by leopal77** [comment]
Following your advice, I reached the application directly from a cluster nodes's ip. The problem is still present so as you stated it's not a load balancer configuration problem.

Regarding the evidence where the communication is lost, from my point of view clients can't be authenticated by the bayeux server. Please inform me which extra technical information you want me to provide.

You can always check [the simple webapp I uploaded](https://github.com/leopal77/CometdAuctionDemo) which is a very simple example reproducing the issue.

Thanks

**Comment by sbordet** [comment]
> The problem is still present

What problem ? The original report was that clients could not connect through the load balancer. If you cannot connect to the server directly, you have a bigger network issue, and that has nothing to do with CometD.

**Comment by leopal77** [comment]
Let me clarify.

My application is a big project containing various sections. On of these sections(where CometD is envolved) is an auction room. 

When **not in cluster mode** everything works as intended(including auction room) even the app is accessed through load balancer or not.

When running the app in cluster mode **with one node active** , auction room is also working great, independently of the way I reach the app(load balancer or active node's ip).

**In cluster mode**, with both nodes active, the auction room **is not** working even I access the app directly through nodes ip or via load balancer. The problem I am reporting is that clients trying to access auction room infinitely exchange `handshake` and `connect` requests with the server. The server always responds to `connect` requests with `402::Unknown Client` and connection is never successful.

In other words, the connectivity problem appears when the oort cloud consists of more than one active node even if I don't use the load balancer.

I hope I made it clear to you.

**Comment by sbordet** [comment]
Sorry but I still don't understand.

When a CometD client connects to a node, it does not matter whether the node is a single node or a node in a Oort cluster. `handshake` and `connect` messages from the client to the server are unrelated to Oort clustering.

There must be something not related to CometD that you have in *cluster mode* that prevents CometD to work properly.

Are you doing some kind of authentication ?
Are you sure the cookies are properly handled ?
Are you sure that the same client always talks to the same server ?

Can you provide the DEBUG logs of the server that is sending the `402::unknown_client` ?

**Comment by leopal77** [comment]
Sorry for my delay.

Finally I 've managed to deal with the issue if have reported.

It was a combination of load balancer's configuration and security restrictions in our application.

This and the [other](https://github.com/cometd/cometd/issues/767) issue can now be closed. 

Thanks for the support Simon.

**Comment by sbordet** [comment]
Thanks for the feedback!
